gnrc_sixlowpan_frag: initial import of minimal forwarding

miri64 opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description
This implements [minimal fragment forwarding](https://datatracker.ietf.org/doc/draft-ietf-6lo-minimal-fragment/) for GNRC.
<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure
There is a test application `tests/gnrc_sixlowpan_frag_minfwd`. Having a set-up with >2 nodes and forward packets between them *should* also work, but for some reason (I suspect the radio) it doesn't work that well at the moment. Once I find out why, I will update the PR and the testing procedures accordingly (see https://github.com/RIOT-OS/RIOT/pull/11256 for what I found out).

`examples/gnrc_networking` should work as normal on a 6LoWPAN compatible board when not using `gnrc_sixlowpan_frag_minfwd` and when using the module both with fragmentation (packet size >100) or without.
<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references
Depends on ~~#11000, #11061, #11063, #11164, #11182, #11263, #11264, and #12629~~ (all merged):

![Route to 6LoWPAN fragment forwarding](http://page.mi.fu-berlin.de/mlenders/sixlo_minfwd.svg)
<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

miri64 commented:
Rebased to current master and #11063  to resolve conflict caused by merging #11135.

miri64 commented:
With the latest addition of #11164 as a dependency and the other fixes I am able to forward, at least when debug output is activated ... I'm not able to forward without it yet though :-(. I will for now first add some test-cases for the bugs I found in the real-world application, and then try to find the bug.

miri64 commented:
OK, I got one successful send (out of 15 so far) :(

miri64 commented:
In https://github.com/RIOT-OS/RIOT/pull/11068#issuecomment-472064281 I wrote:

> With the latest addition of #11164 as a dependency and the other fixes I am able to forward, at least when debug output is activated ... I'm not able to forward without it yet though :-(. I will for now first add some test-cases for the bugs I found in the real-world application, and then try to find the bug.

Some intensive days of hardware debugging later (thank you @PeterKietzmann for all the help and patience) we figured out what happens: appearently most of the time the framebuffer of the driver is overridden with the send data of the forwarded fragment, after the next fragment is received. I'll open a separate issue to explain this in a bit more detail.

miri64 commented:
Rebased and adapted for current master and dependencies (including new dependencies #11263 and #11264)

miri64 commented:
Rebased to include current RC (so I can refer to the 2019.04 release for my experiments)

miri64 commented:
And rebased once again to include fix in `gnrc_netif`

miri64 commented:
Set to WIP because I added some stuff that does not belong with this PR and need to be factored out.

miri64 commented:
Rebased to current master and dependencies.

miri64 commented:
I rebased and squashed this PR to current master and also moved all the unrelated stuff to separate PRs (#11835, #11836, and #11837).

miri64 commented:
Rebased and adapted (#10661) for current dependencies and master

miri64 commented:
Rebased to current master and dependencies since #11000 got merged.

miri64 commented:
Rebased to current master and dependencies.

miri64 commented:
Needs to be adapted for #12300.

miri64 commented:
But for now: rebased and squashed to current #11263 and #11264

miri64 commented:
> Needs to be adapted for #12300.

And #12325.

benpicco commented:
Does this really have a hard dependency on the driver fixes?

miri64 commented:
Hard dependency as in code required that is called here, no. However, minimal fragment forwarding doesn't work without the issue pointed out in #11256 fixed.

miri64 commented:
Rebased and adapted to current master and dependencies. Added #12629 as new dependency.

miri64 commented:
Note to self: as I found a bug in the VRB during the testing of #12303, this PR might also be affected. A rebase after #12848 got merged might be a good idea ;-).

miri64 commented:
Made some modifications to ease the merging with #12303 for my experiments and squashed directly.

miri64 commented:
Some more clean-up work (and making `examples/gnrc_networking` compillable again without MINFWD ^^")

miri64 commented:
Rebased to current master, as #12629 got merged.

miri64 commented:
Rebased to current master to resolve conflicts.

benpicco commented:
This is still waiting for the packet queue, right?

miri64 commented:
Yepp

miri64 commented:
Added some fixes and measures to stabilize the tests in the latest fixup.

miri64 commented:
Rebased, squashed and adapted to current master and dependencies to resolve #11263's conflict with #13065 and this PR's conflicts with #12848 and #13129.

miri64 commented:
Rebased to current master to resolve conflict caused by the merging of #12879.

benpicco commented:
With #12728 being merged now, is this enough to get this working?

miri64 commented:
> With #12728 being merged now, is this enough to get this working?

I still need to do some adaptations to reflect reworks in current master, but in general, yes, with `netdev_ieee802154_submac` this can be merged now under the limitations outlined in [our paper](https://arxiv.org/pdf/1905.08089.pdf) describes.

benpicco commented:
please squash - there hasn't been a review yet.

miri64 commented:
> please squash - there hasn't been a review yet.

Done. Hope the new changes don't get reverted in the review ;-).

miri64 commented:
The draft is [now RFC 8930](https://www.rfc-editor.org/rfc/rfc8930.html), so I updated the doc accordingly

benpicco commented on sys/net/gnrc/network_layer/sixlowpan/frag/minfwd/gnrc_sixlowpan_frag_minfwd.c:
```diff
@@ -0,0 +1,89 @@
+{
+    sixlowpan_frag_t *new;
+    gnrc_pktsnip_t *tmp;
+    const size_t fragsnip_size = sizeof(*frag) -
+                                 ((sixlowpan_frag_1_is((sixlowpan_frag_t *)frag))
+                                 ? sizeof(frag->offset) : 0U);
```
What is `frag->offset` and why is it only stored in the first fragment?

miri64 replied:
It is _not_ stored in the first fragment, that's why it is subtracted. I won't go into 6LoWPAN header format basics here in a comment, but I agree that this code is somewhat unclear and reversed logic. Will fix to
```suggestion
    const size_t fragsnip_size = sizeof(sixlowpan_frag_t) +
                                 ((sixlowpan_frag_1_is((sixlowpan_frag_t *)frag))
                                 ? 0U : sizeof(frag->offset));
```

benpicco commented on sys/net/gnrc/network_layer/sixlowpan/frag/minfwd/gnrc_sixlowpan_frag_minfwd.c:
```diff
@@ -0,0 +1,89 @@
+
+        netif_hdr->flags |= GNRC_NETIF_HDR_FLAGS_MORE_DATA;
+    }
+    LL_PREPEND(pkt, tmp);
```
`gnrc_pkt_prepend()`?

benpicco commented on sys/net/gnrc/network_layer/sixlowpan/gnrc_sixlowpan.c:
```diff
@@ -293,12 +293,36 @@ static void _send(gnrc_pktsnip_t *pkt)
+                fbuf->datagram_size = datagram_size;
+                fbuf->tag = gnrc_sixlowpan_frag_fb_next_tag();
+                fbuf->offset = 0;
+#if IS_USED(MODULE_GNRC_SIXLOWPAN_FRAG_HINT)
```
This is already checked by the encapsulating if block (line 300)

miri64 replied:
True, but the field is only available with `IS_USED(MODULE_GNRC_SIXLOWPAN_FRAG_HINT)` so either we use CPP `#if` for the whole block or just isolated here (or otherwise will get a compiler error, when this module is missing)

miri64 replied:
Will add a comment

benpicco commented on sys/net/gnrc/network_layer/sixlowpan/gnrc_sixlowpan.c:
```diff
@@ -293,12 +293,36 @@ static void _send(gnrc_pktsnip_t *pkt)
+
+        if (IS_USED(MODULE_GNRC_SIXLOWPAN_FRAG_HINT) &&
+            IS_USED(MODULE_GNRC_SIXLOWPAN_FRAG_MINFWD)) {
+            /* prepare for sending with IPHC slack in first fragment */
```
Can we already check if `!ipv6_addr_is_link_local()` here?

miri64 replied:
Mhhh... would be pretty complicated, so I rather would keep this as a later optimization.

benpicco commented on sys/net/gnrc/network_layer/sixlowpan/iphc/gnrc_sixlowpan_iphc.c:
```diff
@@ -1609,9 +1615,21 @@ void gnrc_sixlowpan_iphc_send(gnrc_pktsnip_t *pkt, void *ctx, unsigned page)
+        dst = ipv6_hdr->dst;    /* copying original destination address */
+    }
 
     (void)ctx;
```
guess that's no longer needed

miri64 replied:
Addressed

benpicco commented on sys/net/gnrc/network_layer/sixlowpan/frag/rb/gnrc_sixlowpan_frag_rb.c:
```diff
@@ -592,11 +592,19 @@ static int _rbuf_get(const void *src, size_t src_len,
+            res->pkt = gnrc_pktbuf_add(NULL, NULL, sizeof(ipv6_hdr_t),
+                                       reass_type);
+        }
+        else {
+            res->pkt = gnrc_pktbuf_add(NULL, NULL, 0, reass_type);
+        }
```
What case is this now?

miri64 replied:
Added a comment

benpicco commented on sys/net/gnrc/network_layer/sixlowpan/frag/rb/gnrc_sixlowpan_frag_rb.c:
```diff
@@ -592,11 +592,19 @@ static int _rbuf_get(const void *src, size_t src_len,
+            res->pkt = gnrc_pktbuf_add(NULL, NULL, 0, reass_type);
+        }
+    }
+    else {
+        res->pkt = gnrc_pktbuf_add(NULL, NULL, size, reass_type);
+    }
```
And how does it differ from this one?

miri64 replied:
Added comment.

benpicco commented on sys/net/gnrc/network_layer/sixlowpan/iphc/gnrc_sixlowpan_iphc.c:
```diff
@@ -852,6 +853,7 @@ void gnrc_sixlowpan_iphc_recv(gnrc_pktsnip_t *sixlo, void *rbuf_ptr,
                 }
             }
             if ((ipv6 == NULL) || (res < 0)) {
+                /* TODO fall-back to reassembly in minfwd when ipv6 != NULL */
```
Can you elaborate?

miri64 replied:
It might be, that just a return would be enough here to trigger the mentioned reassembly, need to check.

miri64 replied:
Ah no `ipv6` is changed for sending... I will try to make the comment clearer.

benpicco reviewed (commented):
I'm not familiar with the protocol, so only some nits and questions to better understand the code, where some code comments would also be beneficial for future readers.

Some general questions:
 - how many reassembly buffers can there?  
 - Will we wait a bit before sending the next fragment, so not to immediately cause a collision with the retransmission from the next hop?

I didn't check yet, but I guess there is no effect on code size / behavior if the module is *not* used.

miri64 commented:
> * how many reassembly buffers can there?

Until the RAM is full, I'd say. The number is capped of by `CONFIG_GNRC_SIXLOWPAN_FRAG_RBUF_SIZE`.



> * Will we wait a bit before sending the next fragment, so not to immediately cause a collision with the retransmission from the next hop?

That's basically what is done in SFR (#12648). One could add this as a feature generally to 6LoWPAN, but I rather would do it in a separate PR and when we go into congestion control.



> I didn't check yet, but I guess there is no effect on code size / behavior if the module is _not_ used.

Should be, yes.

miri64 commented on sys/net/gnrc/network_layer/sixlowpan/frag/rb/gnrc_sixlowpan_frag_rb.c:
```diff
@@ -303,6 +350,47 @@ static int _rbuf_add(gnrc_netif_hdr_t *netif_hdr, gnrc_pktsnip_t *pkt,
+                     * fragment in reassembly buffer yet */
+                    sixlowpan_frag_1_is(pkt->data) &&
+                    (entry.super->current_size == frag_size)) {
+                    gnrc_sixlowpan_frag_vrb_t *vrbe;
+                    gnrc_pktsnip_t tmp = {
+                        .data = data,
+                        .size = frag_size,
+                        .users = 1,
+                    };
```
Huh? How? The if above is already 4 spaces to the left.

benpicco replied:
Ah sorry got confused by the indented multi-line condition

miri64 commented on sys/net/gnrc/network_layer/sixlowpan/frag/rb/gnrc_sixlowpan_frag_rb.c:
```diff
@@ -303,6 +350,47 @@ static int _rbuf_add(gnrc_netif_hdr_t *netif_hdr, gnrc_pktsnip_t *pkt,
+                    /* only try minimal forwarding when fragment is the only
+                     * fragment in reassembly buffer yet */
+                    sixlowpan_frag_1_is(pkt->data) &&
+                    (entry.super->current_size == frag_size)) {
```
There would be no fragmentation in that case, so it wouldn't be in the reassembly buffer, so this line would not be called, so no.

miri64 commented on sys/net/gnrc/network_layer/sixlowpan/frag/rb/gnrc_sixlowpan_frag_rb.c:
```diff
@@ -303,6 +350,47 @@ static int _rbuf_add(gnrc_netif_hdr_t *netif_hdr, gnrc_pktsnip_t *pkt,
+            }
+        }
+        if (IS_USED(MODULE_GNRC_SIXLOWPAN_FRAG_MINFWD)) {
+            DEBUG("6lo rbuf: just do normal reassembly\n");
```
As opposed to fragment forwarding.

miri64 replied:
Regarding level of nesting: will try to split this function up in sub-functions.

benpicco replied:
Ah ok - but why is fragment forwarding not performed in this case?

miri64 replied:
Will add a comment. Basically all possible cases for fragment forwarding were checked above, so we fall back to reassembly without this `if (IS_USED(...))` anyway. It's just with `MINFWD` the packet is smaller than it needs to be for reassembly (see `gnrc_pktbuf_add()` stuff you commented on below), so we need to resize it in that case.

miri64 commented on sys/net/gnrc/network_layer/sixlowpan/frag/rb/gnrc_sixlowpan_frag_rb.c:
```diff
@@ -504,7 +592,11 @@ static int _rbuf_get(const void *src, size_t src_len,
         default:
             reass_type = GNRC_NETTYPE_UNDEF;
     }
+#ifdef MODULE_GNRC_SIXLOWPAN_FRAG_VRB
```
I'm a bit confused, why I should use `MODULE_GNRC_NETTYPE_IPV6` here, but will adapt.

miri64 replied:
It is already adapted... I think you commented on an outdated version here.

miri64 commented on sys/net/gnrc/network_layer/sixlowpan/frag/rb/gnrc_sixlowpan_frag_rb.c:
```diff
@@ -642,5 +734,68 @@ int gnrc_sixlowpan_frag_rb_dispatch_when_complete(gnrc_sixlowpan_frag_rb_t *rbuf
     return res;
 }
 
+static bool _check_hdr(gnrc_pktsnip_t *hdr, unsigned page)
```
A person this deep into the 6LoWPAN code will know... It goes through the APIs of every sub-layer. For reference see https://github.com/RIOT-OS/RIOT/issues/8511

benpicco replied:
> A person this deep into the 6LoWPAN code will know…

Those are pretty rare and usually quite busy, so my intention is to soften that requirement ;)

miri64 replied:
So I should comment now every usage of `page`? That seems a bit overwhelming and unrelated to this PR...

benpicco replied:
No, I can trace it back to `gnrc_sixlowpan_frag_rb_add` and there it's documented.

miri64 commented on sys/net/gnrc/network_layer/sixlowpan/frag/rb/gnrc_sixlowpan_frag_rb.c:
```diff
@@ -642,5 +734,68 @@ int gnrc_sixlowpan_frag_rb_dispatch_when_complete(gnrc_sixlowpan_frag_rb_t *rbuf
+            res = -ENOMEM;
+        }
+        else {
+            LL_DELETE(pkt, frag);
```
Will address

miri64 commented:
Pushed everything. Please check, if everything was addressed.

miri64 commented:
Should have at least compile tested before pushing ^^"

benpicco commented on sys/net/gnrc/network_layer/sixlowpan/frag/rb/gnrc_sixlowpan_frag_rb.c:
```diff
@@ -367,30 +372,22 @@ static int _rbuf_add(gnrc_netif_hdr_t *netif_hdr, gnrc_pktsnip_t *pkt,
+             * `_rbuf_get()` */
+            res = _rbuf_resize_for_reassembly(entry.rbuf);
+            if (res = RBUF_ADD_ERROR) {
                 gnrc_pktbuf_release(pkt);
```
`gnrc_pktbuf_release_error()`?

miri64 replied:
Exactly. Even in the reassembly case, this is coming from a different host.

benpicco commented:
Please squash!

miri64 commented:
Squashed!

benpicco commented:
> Squashed!

almost :wink:

miri64 commented:
Oh, there I seemed to have referenced a commit foreign to this PR (used git blame to identify it) Will see where I put it :D

miri64 commented:
Fixed!

benpicco reviewed (approved):
I didn't test multi-hop fragmenting since I don't have a test setup for that (already thought of extending the zep_dispatcher), but I rely on your extensive testing documented in your paper.

I can confirm that there are no regressions when the module is not used, so ACK.

miri64 commented:
Fixed and squashed the one [doxygen error](https://github.com/RIOT-OS/RIOT/pull/11068/checks?check_run_id=1480456214#step:5:46) caused by a legacy name.

benpicco commented:
ah looks like the test will need a `Makefile.ci`.
Do we have a script yet to generate that automatically, like `add_insufficient_memory_board.sh`?

miri64 commented:
And added boards that are to small to `Makefile.ci` to make Murdock happy.

miri64 commented:
Finally :tada:
